Selection Prompt Search

slang25 opened the pull request:
![Multi Selection Search](https://github.com/spectreconsole/spectre.console/assets/1341446/76476b6c-3d9f-4e28-a383-0b8fa047d998)

This PR adds a couple of features to Selection Prompt:
* Search as you type
* Skipping selection of group nodes for `Leaf` selection mode

I've yet to add the tests which is a bit naughty, but wanted to share what I have so far to start some discussion. Is this feature welcome, or is there a better direction?

slang25 commented:
This is somewhat related to #1258, as this is a feature I'd really like in my CLI, however my only feasible options are to fork the library or hope I can get this merged.

martincostello commented on src/Spectre.Console/Extensions/StringExtensions.cs:
```diff
@@ -185,28 +185,35 @@ internal static bool ContainsExact(this string text, string value)
+            return output;
+        }
+
+        foreach (var c in value)
+        {
+            output += mask;
+        }
+
+        return output;
```
Could you make this just:

```csharp
if (mask is null)
{
    return string.Empty;
}

return new string(mask, value.Length);
```

martincostello commented on src/Spectre.Console/Prompts/SelectionPrompt.cs:
```diff
@@ -53,6 +58,11 @@ public sealed class SelectionPrompt<T> : IPrompt<T>, IListPromptStrategy<T>
+    /// <summary>
+    /// Gets or sets a value indicating whether or not the search filter is enabled.
+    /// </summary>
+    public bool SearchFilterEnabled { get; set; } = false;
```
As it's already the default:

```suggestion
    public bool SearchFilterEnabled { get; set; }
```

GOATS2K commented:
This looks incredibly helpful when managing selection prompts with tons of options - I'd love to see this merged soon.

slang25 commented:
Thanks @GOATS2K, I'm pretty happy with it and am using it in one of my apps. I've just updated the API names to be a bit more sensible.

@patriksvensson what are your thoughts on this?

patriksvensson commented on src/Spectre.Console/Prompts/SelectionPrompt.cs:
```diff
@@ -53,6 +58,11 @@ public sealed class SelectionPrompt<T> : IPrompt<T>, IListPromptStrategy<T>
+    /// <summary>
+    /// Gets or sets a value indicating whether or not search is enabled.
+    /// </summary>
+    public bool SearchEnabled { get; set; }
```
We should also add a setting for the text displayed at the bottom of the list `(Type to search)`

patriksvensson commented on src/Spectre.Console/Prompts/List/IListPromptStrategy.cs:
```diff
@@ -7,6 +7,11 @@ namespace Spectre.Console;
+    /// <summary>
+    /// Gets a value indicating whether or not the prompt should skip unselectable items.
+    /// </summary>
+    public bool ShouldSkipUnselectableItems { get; }
```
Can't this be a parameter for `ListPrompt<T>.Show` instead?

patriksvensson commented on src/Spectre.Console/Prompts/SelectionPrompt.cs:
```diff
@@ -93,7 +106,7 @@ public async Task<T> ShowAsync(IAnsiConsole console, CancellationToken cancellat
     ListPromptInputResult IListPromptStrategy<T>.HandleInput(ConsoleKeyInfo key, ListPromptState<T> state)
     {
-        if (key.Key == ConsoleKey.Enter || key.Key == ConsoleKey.Spacebar || key.Key == ConsoleKey.Packet)
+        if (key.Key == ConsoleKey.Enter || key.Key == ConsoleKey.Packet)
```
Why is `ConsoleKey.Spacebar` removed here?

slang25 replied:
This was from testing, I was entertaining the idea of making it configurable. You'd want spacebar to select a multi-selection prompt, but for a regular selection prompt you might want to search for multiple words.

I'll revert it 👍

patriksvensson reviewed (changes requested):
This looks good, but I think some minor things need to be addressed. I would also like to see some tests for the search functionality.

slang25 commented:
@patriksvensson thanks for taking a look, I've addressed the feedback. I've added some tests, I could add some Verify tests too and cover more scenarios but am on the fence, let me know what you think.

sleepyfran commented on src/Spectre.Console/Prompts/SelectionPrompt.cs:
```diff
@@ -169,15 +195,43 @@ IRenderable IListPromptStrategy<T>.Render(IAnsiConsole console, bool scrollable,
 
+            if (searchText.Length > 0 && !(item.Node.IsGroup && Mode == SelectionMode.Leaf))
+            {
+                var index = text.IndexOf(searchText, StringComparison.OrdinalIgnoreCase);
```
This seems to match against markup in the text and causes problems when it matches against it. I took this for a spin on my project and when the line is not current (since that automatically escapes it) it inserts the styles of the search in the middle of other markup, which ends up causing an exception when parsing the markup:
<img width="1178" alt="image" src="https://github.com/spectreconsole/spectre.console/assets/6024783/3f2b02dc-3f73-4bfa-8cdb-7a26acfc84c6">

The original text is `The Hobbit ([bold green3]15.00d$[/])`, if I search for `l` then it matches with the `l` in `bold` (which is markup) and ends up creating `The Hobbit ([bo[bold on yellow]l[/]d green3]15.00d$[/])`, which is invalid markup since styles cannot be nested.

Escaping it seems to work fine, though, so I guess expanding the if in line 193 for search makes sense 😄

slang25 replied:
Thanks, this is a nice find and I'm looking at fixing it now. It's simple enough to discard the markup during the search, but I think the correct behaviour would be to override the style if the found text is within a style (or perhaps in the edge case; spanning across multiple styles), which will be a bit more involved.

Just having a think about it now.

slang25 replied:
I wrote something that would highlight matches across different markup tokens and it was very convoluted and had lots of edge cases. This led me to realise that it's probably best to say that we don't search across tokens.

I've pushed an update that is much simpler, please try again when you get some time 🙂

sleepyfran reviewed (commented):
Thanks a lot for this, really excited to use it! I took it for a spin in a personal fork and found the issue I commented about. Other than that it seems to work perfectly 🙇🏻

slang25 commented:
Hi @patriksvensson, please could you take another look at this PR when you get some time. I'm happy to address any further feedback 🙂

Jacob-32587 commented:
Would love to see this go through soon, extremely useful feature when dealing with large prompts.

patriksvensson commented:
@slang25 I'm rebasing this and fixing the merge conflicts right now

patriksvensson commented:
Merged! Thank you for your contribution. Much appreciated! 👍

slang25 commented:
Thanks @patriksvensson, this library is such a gem for the .NET community, I hope you have time to keep it going 🙂  I was worried we'd lose you to swift 😅

patriksvensson commented:
@slang25 Haha, thanks! I still have a lot of love for dotnet 🙂

fihorvat commented:
@slang25 Sorry for the late comment but it looks like there is a bug. If you try to use something other than `string` for `SelectionPrompt<T>` in combination with the `UseConverter` the cursor selection doesn't work how it should.
Text highlighting works but the cursor doesn't move.
 
![video](https://github.com/user-attachments/assets/af16fdb2-93b8-4cc4-a566-ec43504cb9c9)

Here is a simple test data:

```c#
class TestItem
{
  public string Name { get; set; }
  public string Url { get; set; }
}

var items = new TestItem[] {
  new() { Name = "banana", Url = "something" },
  new() { Name = "orange", Url = "something" },
  new() { Name = "apple", Url = "something" }
};

var prompt = new SelectionPrompt<TestItem>()
  .Title("Select [green]fruit[/]?")
  .PageSize(10)
  .AddChoices(items)
  .UseConverter(x => x.Name)
  .EnableSearch();
prompt.SearchHighlightStyle(Style.Parse("yellow1"));
var response = AnsiConsole.Prompt(prompt);
```

And could you please add the method `SearchHighlightStyle` same as we have for `HighlightStyle`?

patriksvensson commented:
@fihorvat This PR had been merged and no further work is expected by the contributor. Please open a new issue if you experience any bugs.
